File: src/utils/api/version-progress-manager.ts
/**
 * 下载进度模块统一 API 入口
 * `version_progress_manager` IPC 经 `action` 分发；params 字段名 camelCase，6 个 action 均无参数。
 */

import { invoke } from '@tauri-apps/api/core'

/**
 * 调用 version_progress_manager IPC
 * @param action 操作名称（取自 VERSION_PROGRESS_ACTIONS 常量）
 * @param params 参数对象（字段名使用 camelCase），本模块 6 个 action 均无参数
 */
export async function versionProgressManager<T = unknown>(
  action: string,
  params?: unknown,
): Promise<T> {
  return invoke<T>('version_progress_manager', { req: { action, params: params ?? null } })
}

/**
 * 所有可用的 action 名称
 *
 * 与后端 `utils::version_progress_manager::DISPATCHER` 注册的 action 一一对应。
 */
export const VERSION_PROGRESS_ACTIONS = {
  GET_DOWNLOAD_PROGRESS: 'get_download_progress',
  IS_DOWNLOADING: 'is_downloading',
  RESET_DOWNLOAD_PROGRESS: 'reset_download_progress',
  CANCEL_DOWNLOAD: 'cancel_download',
  PAUSE_DOWNLOAD: 'pause_download',
  RESUME_DOWNLOAD: 'resume_download',
} as const

/** action 名称类型 */
export type VersionProgressAction = typeof VERSION_PROGRESS_ACTIONS[keyof typeof VERSION_PROGRESS_ACTIONS]

File: src-tauri/src/utils/version_progress_manager.rs
//! 下载进度统一分发逻辑（version_progress_manager 的工具实现）
//!
//! 使用 `utils::dispatcher::Dispatcher` 注册式分发，6 个 action 均无额外参数仅需 state：
//! `get_download_progress` / `is_downloading` / `reset_download_progress` /
//! `cancel_download` / `pause_download` / `resume_download`。

use once_cell::sync::Lazy;
use tauri::AppHandle;

use crate::commands::version::progress;
use crate::handler;
use crate::state::AppState;
use crate::utils::dispatcher::{ActionRequest, Dispatcher};

static DISPATCHER: Lazy<Dispatcher> = Lazy::new(|| {
    let mut d = Dispatcher::new();

    d.register(
        "get_download_progress",
        handler!(state, _app, _params, {
            let r = progress::get_download_progress(&state).await?;
            serde_json::to_value(r).map_err(|e| e.to_string())
        }),
    );

    d.register(
        "is_downloading",
        handler!(state, _app, _params, {
            let r = progress::is_downloading(&state).await?;
            serde_json::to_value(r).map_err(|e| e.to_string())
        }),
    );

    d.register(
        "reset_download_progress",
        handler!(state, _app, _params, {
            progress::reset_download_progress(&state).await?;
            serde_json::to_value(()).map_err(|e| e.to_string())
        }),
    );

    d.register(
        "cancel_download",
        handler!(state, _app, _params, {
            progress::cancel_download(&state).await?;
            serde_json::to_value(()).map_err(|e| e.to_string())
        }),
    );

    d.register(
        "pause_download",
        handler!(state, _app, _params, {
            progress::pause_download(&state).await?;
            serde_json::to_value(()).map_err(|e| e.to_string())
        }),
    );

    d.register(
        "resume_download",
        handler!(state, _app, _params, {
            progress::resume_download(&state).await?;
            serde_json::to_value(()).map_err(|e| e.to_string())
        }),
    );

    d
});

/// 分发入口
pub async fn dispatch(
    state: AppState,
    app: AppHandle,
    req: ActionRequest,
) -> Result<serde_json::Value, String> {
    DISPATCHER.dispatch(state, app, req).await
}
